Improve SafeHandle management in S.S.C.Csp.

bartonjs opened the pull request:
Correctness: If a SafeKeyHandle is freed after the SafeProvHandle that allocated it, the free method returns false, and ReleaseHandle asserts.  To ensure that doesn't happen, make SafeKeyHandle (and SafeProvHandle) track a parent reference to their related SafeProvHandles.

Performance: SafeProvHandle, SafeKeyHandle, and SafeHashHandle all have InvalidHandle properties which return new objects.  These objects are almost always thrown away immediately, and then get sent to the finalizer queue.  Now they use the SafeHandleCache type for managing singleton InvalidHandle values.

Performance: I doctored up the CSP SafeHandle types to report when they were being finalized, then aggressively sought to get this down to zero finalized handles while running the tests.  The summary of changes:

* Declare the handle allocating methods as out, instead of ref.
* Propagate that change to calling methods
* Dispose any handles that were allocated right before exceptions are thrown
* Factor all of the complicated CspKeyContainerInfo property gets into two
helper methods
    * And ensure that the temporary handles are disposed on exceptional paths
* Delete SafeProvHandle::ReleaseProvider, in the VM it was tracking needing to
free an allocated BStr.
* Change RSACryptoServiceProvider to use lock (_parameters) instead of lock(this)
* Changed _parameters to be readonly, since it could be.
* Move most reads/writes of the SafeHandles in RSACryptoServiceProvider to
properties
    * Ensures proper locking on allocation
    * Ensures proper disposing of previous values on re-assignment

There were also some cleanup elements that snuck in:
* Removed some stale comments
* Corrected some others
* Clipped down the long lines in RSACryptoServiceProvider
* Removed unused using-import statements.

cc: @AtsushiKan @stephentoub @morganbr

Fixes #6211
Fixes #6205
Fixes #6210

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs:
```diff
@@ -38,15 +38,16 @@ internal static partial class CapiHelper
                 {
                     wszUpgrade = MS_ENH_DSS_DH_PROV;
                 }
+                safeProvHandle.Dispose();
```
Rather than having to manually dispose these in a bunch of places, should AcquireCryptContext get an overload or helper that returns handles?

bartonjs replied:
I was very tempted to replace some `int Method(out SafeSomethingHandle, ...)` methods with `SafeSomethingHandle Method(..., out int hr)` so I could consume them in a using statement.

I was trying to keep the changes down a bit (though, yeah, there were a lot); but if anyone wants to second flipping them I'll happily do it :).

morganbr replied:
That's an improvement, but I'd really prefer `SafeSomethingHandle Method()` // no hr, just throws exceptions

bartonjs replied:
Well, in the case of this particular one, that wouldn't really work; since this is doing an opportunistic call.

I agree that there's virtue in that approach, and in the CNG library I'd agree that it was the right thing to do (because that library will continue to evolve).  Here I'm less worried about someone getting things wrong in the future.

I'd be okay with it as a separate commit/PR, but it seems pretty low-priority.
I am, of course, prepared to eat these words in the future :).

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs:
```diff
@@ -69,24 +70,32 @@ public static string UpgradeRSA(int dwProvType, string wszProvider)
             }
-            else if (wszUpgrade == null && requestedBase)
+
+            if (wszUpgrade == null && requestedBase)
```
If I'm understanding this correctly, this code was dead until this change... is there value to upgrading from base to enhanced?

bartonjs replied:
Actually, not only is it likely not worthwhile on Core (since PROV_RSA_AES should always work); I just double-checked the legacy code... and this is also dead code on Desktop.  I'll just delete this clause, entirely.

Good catch.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs:
```diff
@@ -141,48 +150,51 @@ internal static string GetDefaultProvider(int dwType)
+        private static void CreateCSP(CspParameters parameters, bool randomKeyContainer, out SafeProvHandle safeProvHandle)
         {
-            //don't persist in CSP if this is random key container. 
-            safeProvHandle.PersistKeyInCsp = randomKeyContainer ? false : true;
```
Why aren't we propagating this?

bartonjs replied:
We are.  It gets set in OpenCSP.  This one just potentially mutilated the value on an existing SafeHandle before letting it get re-allocated in OpenCSP.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CspKeyContainerInfo.cs:
```diff
@@ -238,15 +188,64 @@ public string UniqueKeyContainerName
+        }
+
+        /// <summary>
+        /// Read a the parameter from the current key using CRYPT_SILENT, to avoid any potential UI prompts.
```
Nit -- "a the parameter"

bartonjs replied:
Oops, fixed.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CspKeyContainerInfo.cs:
```diff
@@ -238,15 +188,64 @@ public string UniqueKeyContainerName
+            const uint SilentFlags = (uint)CapiHelper.CryptAcquireContextFlags.CRYPT_SILENT;
+
+            SafeProvHandle safeProvHandle;
+            int hr = CapiHelper.OpenCSP(_parameters, SilentFlags, out safeProvHandle);
```
Some of the methods you replaced used the verify context. Do we need that for something?

bartonjs replied:
According to MSDN (https://msdn.microsoft.com/en-us/library/windows/desktop/aa379886%28v=vs.85%29.aspx)

> For hardware-based CSPs, such as a smart card CSP, if the pszContainer parameter is NULL or blank, this flag implies that no access to any keys is required, and that no UI should be presented to the user. This form is used to connect to the CSP to query its capabilities but not to actually use its keys.

So it seems like that's the way to talk to the provider to ask it if it's hardware.

morganbr replied:
I understand the silent flag is right, I was also wondering if we need to use  CRYPT_VERIFYCONTEXT in the cases the previous code did.

bartonjs replied:
@morganbr Did I misalign one of the methods?  I see CRYPT_VERIFYCONTEXT methods using ReadDeviceParameterVerifyContext and CRYPT_SILENT methods using ReadKeyParameterSilent, and I don't see anything that used both...

morganbr replied:
Nope, I'm apparently just crazy. Carry on.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CspKeyContainerInfo.cs:
```diff
@@ -51,16 +47,15 @@ public bool Accessible
+
+                if (retVal == null)
                 {
+                    // The key wasn't found, so consider it to be not accessible.
```
I'm not clear on whether this is the right answer -- if we have a provider, it must exist in some capacity. I wonder if this is about ACLs or something.

bartonjs replied:
I -think- it's about AT_EXCHANGE vs AT_SIGNATURE vs AT_SOMETHING_I_DONT_KNOW

bartonjs replied:
Actually, the null means "provider not found".  After that the false would be AT_SIGNATURE or permissions, and true would mean everything's cool.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs:
```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
+            }
+            set
             {
-                lock (this)
```
Do we need to worry about compat issues where user code is locking on the same object?

bartonjs replied:
I can't see any situation where this would be something they were counting on.  But I can always switch it back.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs:
```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
             {
-                GetKeyPair();
+                // Force-read the SafeKeyHandle property, which will summon it into existence.
+                SafeKeyHandle localHandle = SafeKeyHandle;
```
Is this guaranteed to do anything in an optimized build?

stephentoub replied:
The C# compiler won't remove the accessor method call.

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs:
```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
+                        return;
+                    }
+
+                    _safeKeyHandle?.Dispose();
```
Can this race with the getter and end up with the getter returning a disposed key handle? I think we typically save the value into a local and then overwrite the field to prevent that kind of thing (though I'm still not quite sure that's right here)

stephentoub replied:
Yeah, the double-checked locking in this property and in the previous one is fishy.

bartonjs replied:
Is
```C#
set
{
    lock (_parameters)
    {
        SafeKeyHandle current = _safeKeyHandle;

        if (ReferenceEquals(value, current))
        {
            return;
        }

        _safeKeyHandle = value;
        current?.Dispose();
    }
}
```

sufficient?  The get can still have read the old field, but that's also possible with any of the dirty reads that the code was already doing.  It does eliminate the dirty read in the get seeing the disposed value right between dispose and reassignment.

And if we feel more comfortable just unrolling this back to what it used to be, that's fine.  This just seemed as correct and more robust to me.  (As an example, importing a public-only blob overwrote _safeProvHandle, but didn't dispose out the previous one, and it let the stale key handle float for another couple of lines)

morganbr commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs:
```diff
@@ -132,17 +206,7 @@ public bool PersistKeyInCsp
-                    }
-                }
-                return CapiHelper.GetPersistKeyInCsp(_safeProvHandle);
+                return CapiHelper.GetPersistKeyInCsp(SafeProvHandle);
```
I don't think this is thread safe against the provHandle getting replaced. That said, this type isn't documented as thread safe, so why do we have all of this locking?
Arguing with myself here, I'm a little concerned about the possibility of memory safety issues if this code isn't made threadsafe

bartonjs replied:
My take on it is that the locking was to ensure that none of the SafeHandles ever got finalized, so that was the only lock semantic I was preserving.

stephentoub commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs:
```diff
@@ -141,48 +150,51 @@ internal static string GetDefaultProvider(int dwType)
+            int ret = OpenCSP(parameters, dwFlags, out hProv);
             if (S_OK != ret)
             {
+                hProv.Dispose();
```
There's a TODO below this.  Is there a corresponding issue?  Can that TODO be deleted?

bartonjs replied:
@AtsushiKan Was this one yours?  I don't think we need any extra logging here, the exception seems self-sufficient, since it has the hresult.

AtsushiKan replied:
I didn’t add the TODO.

bartonjs replied:
Okay, well; then that means it was created in antiquity, so I'm cool with deleting the logging todos :)

bartonjs replied:
Okay, I've locally removed all of the todos in this project, except 3 - they're now tracked in #6373.

stephentoub commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs:
```diff
@@ -192,35 +204,41 @@ private static void CreateCSP(CspParameters parameters, bool randomKeyContainer,
         {
-            return Interop.CryptAcquireContext(out psafeProvHandle, pszContainer, pszProvider, dwProvType, dwFlags);
+            bool ret = Interop.CryptAcquireContext(out psafeProvHandle, pszContainer, pszProvider, dwProvType, dwFlags);
+            return ret;
```
Just curious, why did you introduce the ret local?

bartonjs replied:
Oops.  There used to be a registration line between them.  (Tracking object creation for the callstack of where things were getting finalized instead of disposed).  I'll revert this.

stephentoub commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs:
```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
+                        _safeKeyHandle?.Dispose();
+                        _safeKeyHandle = null;
+
+                        _safeProvHandle?.Dispose();
```
We just checked that _safeProvHandle isn't null... we don't need to check again.  This can just be _safeProvHandle.Dispose().

bartonjs replied:
Evolution error, good catch.

stephentoub commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SafeCryptoHandles.cs:
```diff
@@ -182,9 +178,34 @@ internal bool PublicOnly
+            _parent = parent;
+
+            bool ignored = false;
+            parent.DangerousAddRef(ref ignored);
```
This probably isn't possible, but just in case, you might want to make this _parent instead of parent.  Otherwise, technically I believe if 'this' were to cease to be rooted elsewhere, it could technically be collected and finalized after setting _parent... then the finalizer would see a non-null parent and call DangerousRelease on it before we added the corresponding ref.  Far fetched, but...

bartonjs replied:
I'm not sure I follow...

_parent = parent;
GC+Finalize => _parent.DangerousRelease() (refcount = 0) + _parent = null;

Option 1)
_parent.DangerousAddRef() => NullReferenceException

Option 2)
parent.DangerousAddRef() => Isn't there an exception when AddReffing from 0 to 1?

Seems equivalently exceptional.  And if the DangerousRelease didn't set it to 0, we'd still want to set it to 1 so it didn't interfere with other children and imbalance them.  So it sounds to me like the parameter's the safer choice... what am I missing?

stephentoub replied:
If you have:
```C#
_parent = parent;
parent.DangerousAddRef(ref ignored);
```
then if `this` is finalized between those two lines, _parent.DangerousReleaseRef will be called with a 0 ref count, causing a failure.

If you instead have:
```C#
_parent = parent;
_parent.DangerousAddRef(ref ignored);
```
then `this` can't be finalized until after the addref because `this` is still rooted.

I don't understand the question about parent.DangerousAddRef in your "Options 2"... can you clarify?  Whether you do ```_parent.DangerousAddRef``` or ```parent.DangerousAddRef``` is immaterial with regards to the state of ```parent```.

bartonjs replied:
Ah, I see, that the pending fieldref would forcibly avoid the situation where the object got collected.  That's what I missed. ("But... that's obvious", well, perhaps, but since `this` getting finalized while an instance method is executing is already non-obvious enough that I was suspending all prejudice)

The exception I was referring to is DangerousAddRef on an IsClosed handle:  https://github.com/dotnet/coreclr/blob/43b39a73cbf832ec13ec29ed356cb75834e7a8d7/src/vm/safehandle.cpp#L112

But that doesn't apply, since the field ref stopped the finalize from happening, which stopped the release from happening.

stephentoub commented on src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SafeCryptoHandles.cs:
```diff
@@ -202,14 +228,41 @@ protected override bool ReleaseHandle()
+            _parent = parent;
+
+            bool ignored = false;
+            parent.DangerousAddRef(ref ignored);
```
Same comment as earlier.

stephentoub commented:
@dotnet-bot test this please

bartonjs commented:
* wiped out the extra todos
* fixed the typo
* collapsed the unnecessary new local
* changed setters to reassign before disposing the previous value
* deleted the unreachable fallback to MS_ENHANCED_PROV
* updated SetParent to use _parent instead of parent as the target of DangerousAddRef

Previous commit was https://github.com/dotnet/corefx/commit/d2eaf416d805eeab999f5ff5427fa886c78a81bc

bartonjs commented:
@dotnet-bot test this please (CI restart)

stephentoub commented:
Test Innerloop OSX Debug Build and Test please
Test Innerloop OSX Release Build and Test please

stephentoub commented:
Test Innerloop OSX Debug Build and Test please (build timed out restoring packages)

bartonjs commented:
@morganbr, @stephentoub : I see two things that are potentially "further changes are expected", but I'm not sure if they're actually in that state or not.

* Undoing the lock target change (`lock (_parameters)` back to `lock (this)`, 'cuz compat.)
* Still not quite happy with the _safeProvHandle/_safeKeyHandle replacement/initialization locking

Since I'm unclear on what action (if any) is expected at this point, figured I'd ask :).

stephentoub commented:
LGTM

stephentoub commented:
I'm going to go ahead and merge this.  If there's additional feedback from others, it can be done as a follow-up PR. Thanks!
